packages/@aws-cdk/assert/lib/assertions/have-resource.ts

```diff
@@ -117,67 +122,89 @@ export interface InspectionFailure {
  *
  * A super-object has the same or more property values, recursing into sub properties if ``allowValueExtension`` is true.
  */
-export function isSuperObject(superObj: any, obj: any, errors: string[] = [], allowValueExtension: boolean = false): boolean {
-  if (obj == null) { return true; }
-  if (Array.isArray(superObj) !== Array.isArray(obj)) {
+export function isSuperObject(superObj: any, pattern: any, errors: string[] = [], allowValueExtension: boolean = false): boolean {
+  if (pattern == null) { return true; }
+  if (Array.isArray(superObj) !== Array.isArray(pattern)) {
     errors.push('Array type mismatch');
     return false;
   }
   if (Array.isArray(superObj)) {
-    if (obj.length !== superObj.length) {
+    if (pattern.length !== superObj.length) {
       errors.push('Array length mismatch');
       return false;
     }
 
     // Do isSuperObject comparison for individual objects
-    for (let i = 0; i < obj.length; i++) {
-      if (!isSuperObject(superObj[i], obj[i], [], allowValueExtension)) {
+    for (let i = 0; i < pattern.length; i++) {
+      if (!isSuperObject(superObj[i], pattern[i], [], allowValueExtension)) {
         errors.push(`Array element ${i} mismatch`);
       }
     }
     return errors.length === 0;
   }
-  if ((typeof superObj === 'object') !== (typeof obj === 'object')) {
+  if ((typeof superObj === 'object') !== (typeof pattern === 'object')) {
     errors.push('Object type mismatch');
     return false;
   }
-  if (typeof obj === 'object') {
-    for (const key of Object.keys(obj)) {
-      if (!(key in superObj)) {
-        errors.push(`Field ${key} missing`);
+  if (typeof pattern === 'object') {
+    for (const [patternKey, patternValue] of Object.entries(pattern)) {
+      if (patternValue === ABSENT) {
+        if (superObj[patternKey] !== undefined) { errors.push(`Field ${patternKey} present, but shouldn't be`); }
```

Comment:
Maybe should use `if (!(patternKey in superObj))` instead of an `undefined` check? Feels like it's a better semantic match?

```js
const obj = { a: undefined, b: 1337 };

obj['a'] !== undefined;
// => false
'a' in obj;
// => true
```

Otherwise I guess the doc of `ABSENT` should probably clarify that it signals the value must be: *absent or explicitly set to `undefined`* (which from the rest of the code seems to be a sufficient condition).

Reply:
Since in the rest of the CDK we don't really make a difference between `undefined` and key absence (they end up the same in the final JSON), I don't artificially want to start making a distinction here. Feels like it would only serve to annoy.